New topic: Soft Fork Activation

harding opened the pull request:
I think this is really putting the `extended` in the `extended_summary` field.  I can probably publish this excessive amount of detail about past soft fork attempts somewhere else if y'all think this should be more focused.

I also think this could maybe do a better job of describing current activation proposals, but they're all still being actively discussed and so I'm not sure it's worth writing about them now only to have to change it in a week or a month.  (Maybe that's why I geeked out about nice, static history.)

amitiuttarwar reviewed (commented):
This is fantastic 💯🙌🏽

Thank you for being an amazing bitcoin historian! Once this is finalized & released, I hope to see a reference to this in an optech newsletter. I had a great time reading it and I'm sure many others will too. 

I left a lot of food for thought, hope some of it is helpful 🙃 

> I'm not sure it's worth writing about them now only to have to change it in a week or a month

I agree with your choice to not flesh them out here.

harding commented on _topics/en/soft-fork-activation.md:
You have a weighted coin that's expected to land on heads 80% of the time, tails 20% of the time.  Imagine two scenarios:

- You flip that coin once every 10 minutes; if after any flip, you saw 950 heads out of the last 1,000, you win.

- Once every 1,000 blocks, you flip that coin 1,000 times.  If you saw 950 or more heads, you win.

In which case is it more probable that you'll win?

harding replied:
So I ran some quick simulations on this point and the effect was much smaller than I thought other people had claimed.  Even assuming crazy long deployments of 10 years (!), I couldn't get a fork to activate a 95% threshold with less than about 90% hashrate actually signaling support.  I personally don't think the difference between 90% and 95% is a big deal, so I'm going to edit out this general point from the article.  Thanks for making me question my assumptions!

```python3
import random

weight = 0.90

tries = 0
while True:
    tries += 1

    signals = 0
    for i in range(1000):
        if random.random() < weight:
          signals += 1
    if signals >= 950:
        print(tries)
        break

    if tries > 52000 * 10:
        print("failed")
        break
```

ajtowns replied:
I think makes a bigger difference if the window is much shorter (like a day or two instead of a week or two). But otherwise I agree -- I've done the same maths repeatedly and got the same result and found people surprised by it. (So it could be worth discussing it just to correct people's misapprehension)

harding commented on _topics/en/soft-fork-activation.md:
BIP148 required *all* blocks signal support for segwit after August 1st.  If the NYA signers just wanted to get segwit activated, they could've tolerated up to 5% non-signaling blocks, and they also could've taken extra time since the BIP141 deployment parameters for BIP9 didn't time out until November 15th.

Instead, BIP91 specified that, once it activated, all non-signaling blocks would be rejected by miners.  It looks pretty clearly to me like they designed their activation of segwit to specifically not violate any of the BIP148 rules so that there was no risk of a consensus failure between UASF nodes and non-UASF nodes (with miners caught in between, probably losing money and getting yelled at by both sides).

harding commented on _topics/en/soft-fork-activation.md:
Yeah.  BIP91 was mainly circulated as a small patch miners could add to their bitcoin source dir.  (Many at the time ran self-patched bitcoinds, although I understand most now run stock Bitcoin Core.)

ajtowns replied:
https://github.com/bitcoin/bitcoin/pull/10900 and https://github.com/bitcoin/bitcoin/pull/10875 were proposed for merging into bitcoin core.

I'm not sure "almost as soon as was possible" is accurate -- the bip's starttime was June 1st, and its window was 2.3 days; not sure what the trigger for miners signalling actually was (release of some fork code?), but it took ~4 periods once signalling started to reach the threshold too -- https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-January/017577.html

harding commented:
Wow @amitiuttarwar, thanks for the comprehensive review!  I applied all the easy code suggestions and answered a few of your questions.  I'll go through and work on the more substantive edits tomorrow.

michaelfolkson commented on _topics/en/soft-fork-activation.md:
The rationale for the MUST_SIGNAL phase (only relevant when LOT=true) is [here](https://bitcoin.stackexchange.com/questions/102585/what-is-the-benefit-of-forced-signaling-in-a-soft-fork-activation-mechanism).

The rationale for no forced signaling during LOCKED_IN phase is [here](https://github.com/bitcoin/bips/pull/1020).

michaelfolkson replied:
Perhaps s/final period/final signaling period

Although this does use the word "signaling" twice in the same sentence.

michaelfolkson replied:
Added a StackExchange question on why BIP 8(LOT=true) doesn't have forced signaling during the LOCKED_IN phase https://bitcoin.stackexchange.com/questions/103755/why-is-there-no-forced-signaling-during-the-locked-in-phase-of-bip-8lot-true

jnewbery replied:
> I don't think you should go into all that here, just saying "final period" isn't accurate

I think "final period" is correct. From the BIP:

> blocks are required to signal in the final period, ensuring the soft fork has locked in by timeoutheight

So "final period" refers to the last possible signaling period, and then there may be a locked_in period, and then active.

michaelfolkson commented on _topics/en/soft-fork-activation.md:
We refer to flag day activation as setting a fixed block height (or time) in the future when the soft fork will activate. The LOT=true aspect of BIP 8 does not set a fixed block height for the activation. It sets a block height for when miners are forced to signal **assuming they have failed to meet the signaling threshold before then**. But of course that forced signaling is not required if they have met the signaling threshold before then.

michaelfolkson replied:
Or in other words the LOT=true aspect doesn't force you to wait for the "flag day". It merely puts an upper limit on when the signaling threshold *must* be met.

jnewbery replied:
I think the wording here is good as it is.

michaelfolkson commented on _topics/en/soft-fork-activation.md:
Perhaps "more sophisticated" and "more robust" introduces some unnecessary opinion. I think working on a "new" activation mechanism in response to challenges encountered is fine wording.

michaelfolkson commented on _topics/en/soft-fork-activation.md:
Agree on "tested" sounding like no real money is on the line. As Amiti says "first used" or "first utilized"?

michaelfolkson commented on _topics/en/soft-fork-activation.md:
BIP 9 was successful in the presence of BIP 148 and BIP 91. I don't know if it can be called "successful" if it needed BIP 148 and BIP 91 to get it safely over the line. Ideally we wouldn't need the use of three BIPs to get a soft fork activated :)

michaelfolkson commented on _topics/en/soft-fork-activation.md:
If "signal BIP148 support for segwit" works just as well as "signal BIP9 support for segwit" I'd suggest dropping the BIP reference and just having "signal support for segwit"

harding commented:
@amitiuttarwar and @michaelfolkson I believe I've addressed all of your feedback.  Thank you both for the detailed comments!

michaelfolkson commented on _topics/en/soft-fork-activation.md:
I think BIP 148 proponents would say there was already a consensus failure and that BIP 148 didn't "create" the risk of a consensus failure. A significant part of the community wanted SegWit to be activated within the BIP 9(1 year) and miners were failing to activate it.

Perhaps instead of "BIP148 created the risk of a consensus failure" replace with "The presence of both BIP 9 and BIP 148 nodes on the network could have led to a consensus failure (or chain split)"

michaelfolkson replied:
You can use "consensus failure" in both a community of humans sense and a network of nodes running software that is potentially incompatible sense. 

When BIP 148 was drafted it was certainly a time where the community of humans disagreed on the best path forward in the face of miners failing to activate SegWit. And this could have led to a consensus failure on the network of nodes. But I don't think it is fair to put all the blame at the feet of the BIP 148 proponents. There wasn't consensus on the best path forward and so the risk of consensus failure was present pre the drafting of BIP 148.

harding replied:
@michaelfolkson 

> You can use "consensus failure" in both a community of humans sense and a network of nodes running software that is potentially incompatible sense.

I've never heard "consensus failure" applied to a community of humans.  None of the first page of results on Google[1] refers to decision making by a community of humans: https://www.google.com/search?q="consensus+failure"  I have heard "failure to reach consensus" and "a loss of consensus" applied to humans, but I think the term "consensus failure" can safely be inferred to only apply to non-sentient computing.

[1] Google personalizes results for the searcher, so my results may be different than yours.  Let me know if that seems to be the case.

> The presence of both BIP 9 and BIP 148 nodes on the network could have led to a consensus failure (or chain split)"

I think you might misunderstand the problem.  Consensus failure was possible between BIP148 nodes and **any** nodes which didn't implement a policy that 100% complied with BIP148 mandatory signaling.  That included BIP141 nodes using BIP9---but it also included nodes that didn't implement segwit (e.g. Bitcoin Core before 0.13.1 and (IIRC) all versions of btcd at the time[1]).

[1] IIRC, btcd didn't add segwit code into its master branch until after segwit was activated.  I don't recall what libbitcoin did.

> I don't think it is fair to put all the blame at the feet of the BIP 148 proponents.

I'm not blaming BIP148 or its proponents, I'm describing one of the fundamental tradeoffs of a soft fork without very widespread enforcement.  There are times when it's more important to risk a consensus failure than it is to do nothing.

> There wasn't consensus on the best path forward and so the risk of consensus failure was present pre the drafting of BIP 148.

I'll edit this to say "[Software implementing] BIP148 created the risk of a consensus failure."

I agree that the seeds of a machine consensus failure were set by the disagreement of people, but the point when that became an actual risk where people could lose money was when someone started running a node with alternative consensus rules.

michaelfolkson replied:
> but the point when that became an actual risk where people could lose money was when someone started running a node with alternative consensus rules.

This argument could apply to anyone ever running BIP 148, UASF or BIP 8(LOT = true) software for a future soft fork. If there is even a single node on the network opposed to forced signaling then no one should run it as by doing so they risk (and are responsible for) consensus failure.

I guess it is technically true. But that position of not wanting to risk consensus failure effectively gives miners a permanent veto on all future soft forks (assuming there is a single node on the network opposed). In that case you have to weigh up the risk of consensus failure with the risk of giving miners a permanent veto of the activation of future soft forks.

michaelfolkson replied:
Let's say BIP 148 and UASF hadn't happened and the BIP 9(1 year) deployment for SegWit had expired and failed. If Core had released something with forced signaling (e.g BIP 149) as a follow up after the BIP 9(1 year) had expired would you say "Core developers have created the risk of a consensus failure"? That is an argument for Core never ever releasing something with forced signaling under **any** circumstances.

harding replied:
@michaelfolkson 

I don't think those situations are analogous.  BIP148 was designed to "cause the existing 'segwit' deployment to activate without needing to release a new deployment."  That implies to me that its proponents weren't sure they had the economic weight necessary to enforce segwit on their own, which further implies that they didn't have the economic weight necessary to prevent consensus failures resulting from their forced signaling requirement.  In short: they were taking a gamble, and they knew it.

If Bitcoin Core ever releases a flag day or mandatory signaling soft fork that seems unlikely to gain overwhelming adoption among economic nodes before it activates (and is not an emergency), I would certainly also describe that as having created the risk of a consensus failure.  However, I believe it is possible for that project to wait for a change to have demonstrated overwhelming support and then release a flag day activation far enough in the future that, combined with almost all economic nodes already running Bitcoin Core, indicates almost everyone will be running that code by the time it activates.

(I think I've spent too much time on this argument already.  I will gladly read any response you send and try to incorporate your points into my edits, but I do not expect to respond further.  Thank you for your feedback.)

michaelfolkson replied:
The wording is hard, I suspect whatever wording you use someone will find fault with it. I won't take up any more of your time. Thanks for your responses.

jnewbery commented on _topics/en/soft-fork-activation.md:
```suggestion
  - title: "OP_CHECKTEMPLATEVERIFY workshop discussion about activation mechanisms"
```

jnewbery commented on _topics/en/soft-fork-activation.md:
```suggestion
  - title: "BIPs #1020 updates BIP8 to no longer require signaling during lockin period"
```

jnewbery reviewed (commented):
Great stuff, @harding! Thank you so much for documenting this.

michaelfolkson commented on _topics/en/soft-fork-activation.md:
It is in reference to this [mailing list post](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-February/018498.html).

Luke's position (afaiu) is that if it was up to him lockinontimeout would always be true. It wouldn't be a parameter that can be set to true or false. However, he merged the parameter into revised BIP 8 and he understands that some people have the opposing view, that the lockinontimeout parameter should always be false. And other people think it should sometimes be true and sometimes be false. 

Hence I wouldn't call Luke a detractor of BIP 8. He is a detractor of some nodes on the network running lockinontimeout=true whilst other nodes are running lockinontimeout=false.

So in short, I would say there are detractors for including a lockinontimeout parameter in BIP 8. But I think all the other parameters in BIP 8 don't have detractors and everyone appreciates the flexibility that having multiple parameters brings when discussing future soft fork activations.

jnewbery replied:
I'd suggest just removing the words "(including BIP8's own champion)".

ajtowns commented on _topics/en/soft-fork-activation.md:
Making blocks faster to propogate and validate? Could use some supporting links...

ajtowns commented on _topics/en/soft-fork-activation.md:
Unnecessar==il==y? Penalisation of?

ajtowns commented on _topics/en/soft-fork-activation.md:
Technically it expired when BIP141 reached locked in.

harding commented:
Just a note that I'm still working on making edits for the above feedback.  I was optimistically hoping to finish my edits on Friday and again today; now I'm hoping to finish tomorrow.

harding commented on _topics/en/soft-fork-activation.md:
Modern GitHub's "editted" tags have spoiled you.  :-)  That PR's original description was almost certainly edited, as can be seen from the discussion and from the date stamp in the actual commit (which is days after the PR was opened, although I just now realized GitHub doesn't show that date---I think it shows the date the commit was added to master, i.e. the merge date).

I'll edit the text to try to make the sequence of events clearer and better supported by the links.  (Note, the reason I go into this detail is that the two activation thresholds always seemed odd to me; I think it makes more sense when they're understood as a proposal on top of a previous proposal.)

jnewbery replied:
Ah thank you! That makes total sense. The way Github's edit feature is used to change history is infuriating, and makes life much more difficult for code archaeologists.

jnewbery replied:
Thanks for the additional references!

jnewbery commented:
Oh! I didn't notice the most recent push here. @harding is this ready for re-review?

harding commented:
@jnewbery 

> @harding is this ready for re-review?

Indeed it is.  I believe all feedback to date has been addressed.  Sorry I didn't say anything.

jnewbery commented on _topics/en/soft-fork-activation.md:
```diff
@@ -0,0 +1,462 @@
+  31,000, which occurred on 22 December 2009.  This *hardcoded height*
+  activation mechanism was used for at least one [other][block size sf]
+  early soft fork when most development was done by Satoshi Nakamoto.
+
```
added!

jnewbery commented on _topics/en/soft-fork-activation.md:
Marking resolved.

jnewbery commented on _topics/en/soft-fork-activation.md:
s/64-bit floating-point value/64-bit integer/

jnewbery commented on _topics/en/soft-fork-activation.md:
Should this be 92,233,720,368.54775807 BTC / 9,223,372,036,854,775,807 sats (the largest possible signed 64-bit integer)?

jnewbery commented on _topics/en/soft-fork-activation.md:
Was it an intentional decision to refer to an "old version of Bitcoin" and and "emergency release of Bitcoin **Core**"? I believe the name Bitcoin Core was only used for version 0.9 onwards.

harding replied:
Huh, it was deliberate because I was sure that was in 0.8.0, but the 0.9.0 release notes do indeed contain the rebranding notice.  Thanks for reading so closely!

jnewbery commented:
The new section on **Emergency activations** is great! I've left some minor review comments on the section.

Additionally, there are two previous review comments that I think still need to be addressed:

- https://github.com/bitcoinops/bitcoinops.github.io/pull/531#discussion_r599592479
- https://github.com/bitcoinops/bitcoinops.github.io/pull/531#discussion_r600429688

luke-jr commented on _topics/en/soft-fork-activation.md:
This is nonsense. The risk of consensus failure was strictly limited to non-upgraded nodes, and no different than the risk with any other softfork.

luke-jr commented on _topics/en/soft-fork-activation.md:
Eh? BIP8 was a generalised BIP148. Nothing to do with BIP149.

luke-jr commented on _topics/en/soft-fork-activation.md:
No, NYA/BIP91 were several months later, well after it became clear BIP148 would succeed. Its only purpose was face saving and attempting to tie a hardfork to it.

luke-jr commented on _topics/en/soft-fork-activation.md:
It's confusing to pretend BIP91 was an actual softfork that activated. No economic full nodes ever enforced it. At a technical level, it was merely a 51% attack. The only reason it was "okay" at all, was because it was a form of compliance with BIP148.

harding replied:
I would personally prefer to pretend that BIP91 never existed.  I added this text about it in response to another reviewer reasonably wanting to know what role BIP91 played in taproot's activation.  I think this text is reasonably clear about what happened, saying (emphasis added): "**Miners** used a BIP9 versionbit to signal whether they would enforce a temporary rule [...]"

luke-jr commented on _topics/en/soft-fork-activation.md:
Seems like the only "debate" against BIP148's success is ego-centric refusal of some opponents to admit they were wrong. Late 2017, it was widely acknowledged, even by people who didn't believe in BIP148, that it had succeeded.

harding replied:
I strongly agree that some people really do seem to be egocentrically invested in pushing a particular narrative.

harding commented:
Pushed an edit that I hope fully addresess @jnewbery and @ajtowns feedback.  Thanks (and double thanks to John for pointing me at feedback from AJ that I had somehow not seen).

The same edit also addresses all comments from @luke-jr except two where I left a specific reply.  Thanks especially for [this comment](https://github.com/bitcoinops/bitcoinops.github.io/pull/531#discussion_r607992344) where I had indeed misconstrued the sequence of events.

If this is looking pretty good, I'd like to squash it tomorrow or Thursday.

jnewbery commented:
This is ready for merge. Fantastic work as always, @harding! Thank you for documenting this.
